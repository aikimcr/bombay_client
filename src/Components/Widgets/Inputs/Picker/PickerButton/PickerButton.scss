@use '@styles/mixins.scss';

.picker-button {
  display: flex;
  flex-direction: row;
  gap: .5rem;
  justify-content: start;
  align-items: center;
  anchor-name: --picker-button;

  @include mixins.buttonBase($borderRadius: 1rem, $borderStyle: solid, $borderWidth: 1px);

  &:hover:not(:disabled) {
    box-shadow: var(--color-white-200) 2px 2px 1px 1px
  }

  &:active:not(:disabled) {
    box-shadow: var(--color-white-200) 2px 2px 1px 1px inset
  }
}