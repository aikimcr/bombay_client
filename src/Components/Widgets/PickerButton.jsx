import PropTypes from 'prop-types';
import { useRef, useState } from 'react';
import { PickerList } from '.';
import { Button } from './Buttons';

import './OldLabeledInput.scss';
function PickerButton(props) {
  const topRef = useRef(null);
  const inputRef = useRef(null);

  const modelId = props.model ? props.model.id : 'isNew';
  const buttonId = `${props.modelName}-name-${modelId}`;

  const [showPickerList, setShowPickerList] = useState(false);
  const [currentModel, setCurrentModel] = useState(
    props.model ? props.model : null,
  );

  const buttonLabel = currentModel ? currentModel.name : '<Please Choose>';

  const initialValue = currentModel ? currentModel.id : 0;

  function modelPicked(newModel) {
    setCurrentModel(newModel);
    inputRef.current.value = newModel[props.fieldName];
    props.onModelPicked(newModel);
    setShowPickerList(false);
  }

  function pickAModel(evt) {
    evt.preventDefault();
    setShowPickerList((oldShow) => !oldShow);
  }

  return (
    <div
      ref={topRef}
      className="picker-button"
      data-modelname={props.modelName}
      data-targetfield={props.targetField}
    >
      <label htmlFor={buttonId}>{props.labelText}</label>
      <input
        type="number"
        ref={inputRef}
        name={props.targetField}
        defaultValue={initialValue}
      ></input>

      <Button
        id={buttonId}
        text={buttonLabel}
        disabled={false}
        onClick={pickAModel}
      />

      <PickerList
        pickModel={modelPicked}
        isOpen={showPickerList}
        collectionClass={props.collectionClass}
      />
    </div>
  );
}

PickerButton.propTypes = {
  collectionClass: PropTypes.func.isRequired, // To get the list of models
  model: PropTypes.object, // A currently set model (optional)
  modelName: PropTypes.string.isRequired, // Should match collectionClass, mostly a convenience
  fieldName: PropTypes.string.isRequired, // The field from the model used to set the target value
  targetField: PropTypes.string.isRequired, // The key in the data from the form
  labelText: PropTypes.string.isRequired,
  onModelPicked: PropTypes.func.isRequired, // Called with the selected model from the collection
};

export default PickerButton;
