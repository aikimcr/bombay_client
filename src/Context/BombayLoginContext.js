import { createContext } from "react";

const BombayLoginContext = createContext({
  loggedIn: false,
  showLoginForm: false,
  setLoggedIn: null,
  setShowLogin: null,
});

export default BombayLoginContext;
