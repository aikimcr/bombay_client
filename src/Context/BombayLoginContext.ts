import { createContext, Dispatch } from "react";

interface BombayLoginContextValues {
  loggedIn: boolean;
  setLoggedIn?: Dispatch<boolean>;
}

export const BombayLoginContext = createContext({
  loggedIn: false,
  setLoggedIn: null,
});

export default BombayLoginContext;
