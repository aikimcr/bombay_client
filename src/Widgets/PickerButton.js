import PropTypes from 'prop-types';
import { useRef, useState } from 'react';
import PickerList from './PickerList';
import Button from './Button';

import './LabeledInput.scss';

function PickerButton(props) {
    const topRef = useRef(null);

    const modelId = props.model ? props.model.get('id') : 'isNew';
    const buttonId = `${props.modelName}-name-${modelId}`;

    const [showPickerList, setShowPickerList] = useState(false);
    const [currentModel, setCurrentModel] = useState(props.model ? props.model : null);
    
    const buttonLabel = currentModel ?
                        currentModel.get('name') :
                        '<Please Choose>';

    function modelPicked(newModel) {
        setCurrentModel(newModel);
        props.onModelPicked(newModel);
        setShowPickerList(false);
    }

    function pickAModel(evt) {
        evt.preventDefault();
        setShowPickerList(true);
    }

    return (
        <div 
            ref={topRef}
            className='picker-button'
            data-modelname={props.modelName}
        >
            <label htmlFor={buttonId}>{props.labelText}</label>
            <Button
                id={buttonId}
                label={buttonLabel}
                disabled={false}
                onClick={pickAModel}
            />
            <PickerList
                pickModel={modelPicked}
                isOpen={showPickerList}
                collectionClass={props.collectionClass}
            />
        </div>
    );
}

PickerButton.propTypes = {
    modelName: PropTypes.string.isRequired,
    labelText: PropTypes.string.isRequired,
    collectionClass: PropTypes.func.isRequired,
    onModelPicked: PropTypes.func.isRequired,
    model: PropTypes.object,
}

export default PickerButton;