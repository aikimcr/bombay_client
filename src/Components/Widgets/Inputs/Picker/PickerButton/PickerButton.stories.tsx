import type { Meta, StoryObj } from '@storybook/react-webpack5';
import React, { useState } from 'react';
import { fn } from 'storybook/test';

import { PickerButton } from './PickerButton';

const meta = {
  component: PickerButton,

  tags: ['autodocs'],

  args: {
    openStateToggle: fn(),
  },
} satisfies Meta<typeof PickerButton>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    id: 'pick-button',
    buttonText: 'Pick One',
    openState: false,
    popoverTarget: 'pop-target',
  },
  render: (args) => {
    const [openState, setOpenState] = useState(false);

    const openStateToggle = () => {
      setOpenState((oldState) => !oldState);
    };

    return (
      <PickerButton
        {...args}
        openState={openState}
        openStateToggle={openStateToggle}
      />
    );
  },
};

export const Disabled: Story = {
  args: {
    id: 'pick-button',
    buttonText: 'Pick One',
    openState: false,
    popoverTarget: 'pop-target',
  },
  render: (args) => {
    return <PickerButton {...args} disabled={true} />;
  },
};
