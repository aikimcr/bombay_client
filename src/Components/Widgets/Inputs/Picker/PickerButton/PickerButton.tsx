import React from 'react';

import './PickerButton.scss';

interface PickerButtonProps {
  id?: string;
  buttonText: string;
  disabled?: boolean;
  openState: boolean;
  openStateToggle: () => void;
  popoverTarget: string;
}

export const PickerButton: React.FC<PickerButtonProps> = ({
  id = 'picker-button',
  buttonText,
  disabled = false,
  openState,
  openStateToggle,
  popoverTarget,
}): React.ReactElement => {
  return (
    <button
      id={id}
      className="picker-button"
      data-testid="picker-button"
      type="button"
      onClick={openStateToggle}
      aria-haspopup="menu"
      aria-expanded={openState}
      aria-controls={popoverTarget}
      popoverTarget={popoverTarget}
      popoverTargetAction="toggle"
      disabled={disabled}
    >
      {buttonText}
      {openState ? (
        <i className="fa-solid fa-caret-up"></i>
      ) : (
        <i className="fa-solid fa-caret-down"></i>
      )}
    </button>
  );
};
