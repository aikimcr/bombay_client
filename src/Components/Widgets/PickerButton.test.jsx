import { useState } from "react";
import { act, fireEvent, render, screen } from "@testing-library/react";

import {
  mockGetFromURLString,
  mockLogin,
  mockLoginStatus,
  mockLogout,
  mockPostToURLString,
  mockPrepareURLFromArgs,
  mockPutToURLString,
  mockRefreshToken,
  mockServerBasePath,
  mockServerHost,
  mockServerPort,
  mockServerProtocol,
} from "../../Network/testing";

jest.mock("../../Network/Login", () => {
  const originalModule = jest.requireActual("../../Network/Login");

  return {
    __esModule: true,
    ...originalModule,
    loginStatus: mockLoginStatus,
    refreshToken: mockRefreshToken,
    login: mockLogin,
    logout: mockLogout,
  };
});

jest.mock("../../Network/Network", () => {
  const originalModule = jest.requireActual("../../Network/Network");

  return {
    __esModule: true,
    ...originalModule,
    serverProtocol: mockServerProtocol,
    serverHost: mockServerHost,
    serverBasePath: mockServerBasePath,
    serverPort: mockServerPort,
    prepareURLFromArgs: mockPrepareURLFromArgs,
    getFromURLString: mockGetFromURLString,
    postToURLString: mockPostToURLString,
    putToURLString: mockPutToURLString,
  };
});

import {
  MockTestCollection,
  mockFetchBody,
  mockModels,
  MockPickerList,
} from "./testing";

jest.mock(".", () => {
  const originalModule = jest.requireActual(".");

  return {
    __esModule: true,
    ...originalModule,
    PickerList: MockPickerList,
  };
});

import BombayLoginContext from "../../Context/BombayLoginContext";

import PickerButton from "./PickerButton.jsx";

let testSetLoggedIn;

function TestWrapper({ loggedIn, children }) {
  const [loginState, setLoginState] = useState({ loggedIn });

  testSetLoggedIn = (isLoggedIn) => {
    setLoginState((oldState) => {
      return { ...oldState, loggedIn: isLoggedIn };
    });
  };

  return (
    <BombayLoginContext.Provider value={loginState}>
      {children}
    </BombayLoginContext.Provider>
  );
}

jest.useFakeTimers();

const testToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJGREM4MTEzOCIsInVzZXIiOnsiaWQiOjEsIm5hbWUiOiJhZG1pbiIsImFkbWluIjpmYWxzZX0sImlhdCI6MTY2NTk2NTA5OX0.2vz14X7Tm-oFlyOa7dcAF-5y5ympi_UlWyJNxO4xyS4";

function setupLogin(loggedIn = true, token = testToken) {
  const loginPromise = PromiseWithResolvers();
  loginPromise.resolve({ loggedIn, token });
}

beforeEach(() => {
  localStorage.setItem("jwttoken", testToken);
});

afterEach(() => {
  localStorage.removeItem("jwttoken");
});

it("Component should match snapshot", async () => {
  setupLogin();

  const getPromise = PromiseWithResolvers();
  mockGetFromURLString.mockReturnValue(getPromise.promise);

  const onModelPicked = jest.fn();

  const { asFragment } = render(
    <TestWrapper loggedIn={true}>
      <PickerButton
        collectionClass={MockTestCollection}
        modelName="table1"
        fieldName="id"
        targetField="table2_id"
        labelText="Pick A Model"
        onModelPicked={onModelPicked}
      />
    </TestWrapper>,
  );

  await act(async () => {
    getPromise.resolve(mockFetchBody);
  });

  expect(asFragment()).toMatchSnapshot();
});

it("should hide the list on load", async () => {
  setupLogin();

  const getPromise = PromiseWithResolvers();
  mockGetFromURLString.mockReturnValue(getPromise.promise);

  const onModelPicked = jest.fn();

  render(
    <TestWrapper loggedIn={true}>
      <PickerButton
        collectionClass={MockTestCollection}
        modelName="table1"
        fieldName="id"
        targetField="table2_id"
        labelText="Pick A Model"
        onModelPicked={onModelPicked}
      />
    </TestWrapper>,
  );

  await act(async () => {
    getPromise.resolve(mockFetchBody);
  });

  expect(screen.queryByTestId("mock-picker-list")).toBeNull();
});

it("should show the list on click", async () => {
  setupLogin();

  const getPromise = PromiseWithResolvers();
  mockGetFromURLString.mockReturnValue(getPromise.promise);

  const onModelPicked = jest.fn();

  render(
    <TestWrapper loggedIn={true}>
      <PickerButton
        collectionClass={MockTestCollection}
        modelName="table1"
        fieldName="id"
        targetField="table2_id"
        labelText="Pick A Model"
        onModelPicked={onModelPicked}
      />
    </TestWrapper>,
  );

  await act(async () => {
    getPromise.resolve(mockFetchBody);
  });

  let listComponent = screen.queryByTestId("mock-picker-list");
  expect(listComponent).toBeNull;

  const showButton = screen.getByText(/Please Choose/);

  await act(async () => {
    fireEvent.click(showButton);
  });

  listComponent = screen.getByTestId("mock-picker-list");
  expect(listComponent).toBeInTheDocument();
});

it("should call the callback and close the list when an item is clicked", async () => {
  setupLogin();

  const getPromise = PromiseWithResolvers();
  mockGetFromURLString.mockReturnValue(getPromise.promise);

  const onModelPicked = jest.fn();

  render(
    <TestWrapper loggedIn={true}>
      <PickerButton
        collectionClass={MockTestCollection}
        modelName="table1"
        fieldName="id"
        targetField="table2_id"
        labelText="Pick A Model"
        onModelPicked={onModelPicked}
      />
    </TestWrapper>,
  );

  await act(async () => {
    getPromise.resolve(mockFetchBody);
  });

  const showButton = screen.getByText(/Please Choose/);

  await act(async () => {
    fireEvent.click(showButton);
  });

  expect(onModelPicked).not.toBeCalled();

  const el = screen.getByText("Pick Model");

  await act(async () => {
    fireEvent.click(el);
  });

  expect(onModelPicked).toBeCalled();
  expect(onModelPicked).toBeCalledWith(mockModels[2]);

  const listComponent = screen.queryByTestId("mock-picker-list");
  expect(listComponent).toBeNull;
});

it("should close the list without calling if the button is pushed again", async () => {
  setupLogin();

  const getPromise = PromiseWithResolvers();
  mockGetFromURLString.mockReturnValue(getPromise.promise);

  const onModelPicked = jest.fn();

  render(
    <PickerButton
      collectionClass={MockTestCollection}
      modelName="table1"
      targetField="table2_id"
      fieldName="table2_id"
      labelText="Pick A Model"
      onModelPicked={onModelPicked}
    />,
  );

  await act(async () => {
    getPromise.resolve(mockFetchBody);
  });

  const showButton = screen.queryByText(/Please Choose/);

  await act(async () => {
    fireEvent.click(showButton);
  });

  expect(onModelPicked).not.toBeCalled();

  let listComponent = screen.getByTestId("mock-picker-list");
  expect(listComponent).toBeInTheDocument();

  expect(onModelPicked).not.toBeCalled();

  await act(async () => {
    showButton.click();
  });

  listComponent = screen.queryByTestId("mock-picker-list");
  expect(listComponent).not.toBeInTheDocument();
});
