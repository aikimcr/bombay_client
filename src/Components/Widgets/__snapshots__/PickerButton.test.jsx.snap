// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="picker-button"
    data-modelname="table1"
    data-targetfield="table2_id"
  >
    <label
      for="table1-name-isNew"
    >
      Pick A Model
    </label>
    <input
      name="table2_id"
      type="number"
      value="0"
    />
    <button
      class="button button__primary"
      id="table1-name-isNew"
    >
      &lt;Please Choose&gt;
    </button>
  </div>
</DocumentFragment>
`;
